File: src/python/modules/multistart/greedy_based_heuristics/greedy_heuristic.py
import math

from modules.data import InstanceData
from modules.solutions import Solution, print_detail_melo_formulation_solution
from modules.parameters import ParameterData

from .init_solution import init_solution, tightest_time_windows
from .check_insertion import check_insertion
from .entities_greedy_heur import InsertionData
from .update_solution import update_solution, update_machines_indexes


def get_insertion_with_less_increase_in_comp_time(
    inst: InstanceData, sol: Solution, pJob: int, dJob: int
) -> InsertionData:
    best_ins_data = InsertionData(
        is_feasible=False,
        cost=math.inf,
        pPos=0,
        dPos=0,
        pJob=0,
        dJob=0,
        k=0,
        machine_travels=[],
    )

    for k in inst.K:
        vehicle_route = sol.vehicles[k]
        for pPos in range(1, len(vehicle_route)):
            for dPos in range(pPos, len(vehicle_route)):
                print("Analyzing: ", k, " ", pPos, " ", dPos, " ", pJob, " ", dJob, sep="")
                check_ins_data = check_insertion(sol, k, pPos, dPos, pJob, dJob, inst)
                print(check_ins_data)
                if check_ins_data.feasible:
                    if check_ins_data.cost < best_ins_data.cost:
                        best_ins_data = InsertionData(
                            is_feasible=check_ins_data.feasible,
                            cost=check_ins_data.cost,
                            pPos=pPos,
                            dPos=dPos,
                            pJob=pJob,
                            dJob=dJob,
                            k=k,
                            machine_travels=check_ins_data.possibleMachineTravels,
                        )
    
    print()
    print()
    print()
    print()
    print()
    print()
    return best_ins_data

# ! Not working yet
def greedy_heuristic(inst: InstanceData, params: ParameterData) -> Solution:
    sol = init_solution(inst)
    non_serviced_reqs = tightest_time_windows(inst).copy()
    idx_req_to_serve = 0
    req_not_inserted = False

    print([non_serviced_reqs[i] + 1 for i in range(len(non_serviced_reqs))])

    while idx_req_to_serve < len(non_serviced_reqs) and not req_not_inserted:
        pJob = non_serviced_reqs[idx_req_to_serve]
        dJob = pJob + inst.n
        print(pJob, dJob)

        best_ins_data = get_insertion_with_less_increase_in_comp_time(
            inst, sol, pJob, dJob
        )

        if best_ins_data.is_feasible:
            print()
            sol = update_solution(sol, best_ins_data, inst)
            print("PRINT ROUTES")
            for k in inst.K:
                print("k:", k, "-", sol.vehicles[k][-1].servST)
            print()
            
            if idx_req_to_serve == 10:
                break
                exit(0)
        else:
            req_not_inserted = True
        print()
        print("Solution value: ",sol.value)
        idx_req_to_serve += 1

    update_machines_indexes(sol)
    sol.is_feasible = not req_not_inserted
    print_detail_melo_formulation_solution(inst, sol)
    exit(0)

    # Optional validation
    # if validateSolution(inst, sol, params):
    #     sol.feasible = True
    # else:
    #     sol.feasible = False

    return sol

File: src/julia/modules/Multistart/src/greedy_heuristic.jl
function getInsertionWithLessIncreaseInCompTime(
	inst::InstanceData,
	sol::Solution,
	p_job::Int64,
	d_job::Int64,
)::InsertionData
	best_ins_data = InsertionData(false, Inf64, 0, 0, 0, 0, 0, PossibleMachineTravel[])
	for k in inst.K
		for p_pos in 2:length(sol.vehicles[k])
			for d_pos in p_pos:length(sol.vehicles[k])
				check_ins_data = checkInsertion(sol, k, p_pos, d_pos, p_job, d_job, inst)
				if check_ins_data.feasible
					if check_ins_data.cost < best_ins_data.cost
						best_ins_data = InsertionData(
							check_ins_data.feasible,
							check_ins_data.cost,
							p_pos,
							d_pos,
							p_job,
							d_job,
							k,
							check_ins_data.possibleMachineTravels,
						)
					end
				end
			end
		end
	end
	return best_ins_data
end # function getInsertionWithLessIncreaseInCompTime()

function greedyHeuristic(inst::InstanceData, params::ParameterData)::Solution
	sol = initSolution(inst)
	nonServicedReqs = copy(tightestTimeWindows(inst))
	idxReqToServe = 1
	reqNotInserted = false
	while idxReqToServe <= length(nonServicedReqs)
		p_job = nonServicedReqs[idxReqToServe]
		d_job = p_job + inst.n

		best_ins_data = getInsertionWithLessIncreaseInCompTime(inst, sol, p_job, d_job)

		if best_ins_data.feasible
			sol = updateSolution(sol, best_ins_data, inst)
		else
			reqNotInserted = true
			idxReqToServe = length(nonServicedReqs)
		end
		idxReqToServe += 1
	end

	updateMachinesIndexes(sol)

	sol.feasible = !reqNotInserted

	# print_timeline_solution(inst, sol)
	# if validate_solution(inst, sol, params)
	# 	sol.feasible = true
	# else
	# 	sol.feasible = false
	# end
	return sol
end # function greedyHeuristic()
